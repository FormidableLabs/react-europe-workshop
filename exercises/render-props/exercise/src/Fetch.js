import React, { Component } from 'react';

class Fetch extends Component {
  render() {
    return <p>Fetch Component</p>;
  }
}

export default Fetch;
