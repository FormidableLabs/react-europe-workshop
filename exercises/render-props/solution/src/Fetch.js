import React, { Component } from 'react';

class Fetch extends Component {
  state = {
    loading: true,
    error: null,
    data: null,
  };
  componentDidMount() {
    fetch(this.props.url)
      .then(res => res.json())
      .then(data => {
        this.setState({
          data,
          loading: false,
        });
      })
      .catch(err => {
        this.setState({
          error: err,
          loading: false,
        });
      });
  }
  render() {
    return this.props.children(this.state);
  }
}

export default Fetch;
