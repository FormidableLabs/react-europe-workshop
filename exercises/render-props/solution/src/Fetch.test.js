import React from 'react';
import renderer from 'react-test-renderer';
import Fetch from './Fetch';

global.fetch = jest.fn();

describe('Fetch Component', () => {
  global.fetch.mockReturnValue(
    new Promise(resolve => {
      setTimeout(() => {
        resolve({
          json: () => Promise.resolve('HEY'),
        });
      }, 200);
    })
  );
  it('renders a snapshot that is good', () => {
    const tree = renderer
      .create(
        <Fetch url="http://google.com">
          {args => <p style={{ color: !args.loading ? 'blue' : 'red' }}>YAY</p>}
        </Fetch>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('starts with loading equals true', () => {
    let result;
    const tree = renderer.create(
      <Fetch url="whatever">
        {({ loading }) => {
          result = loading;
          return null;
        }}
      </Fetch>
    );

    expect(result).toBe(true);
  });

  it('should do data', done => {
    let result;
    const tree = renderer.create(
      <Fetch url="whatever">
        {data => {
          console.log(data);
          result = data;
          return null;
        }}
      </Fetch>
    );

    expect(result.loading).toBe(true);
    setTimeout(() => {
      expect(result.data).toBe('HEY');
      done();
    }, 300);
  });

  it('should surface request errors', done => {
    global.fetch.mockReturnValue(
      new Promise((resolve, reject) => {
        setTimeout(() => {
          reject('Noooooooo');
        }, 200);
      })
    );

    let result;

    const tree = renderer.create(
      <Fetch url="whatever">
        {data => {
          console.log(data);
          result = data;
          return null;
        }}
      </Fetch>
    );

    setTimeout(() => {
      expect(result.error).toBe('Noooooooo');
      done();
    }, 300);
  });
});
