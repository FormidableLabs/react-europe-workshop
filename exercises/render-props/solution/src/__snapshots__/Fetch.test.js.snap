// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch Component renders a snapshot that is good 1`] = `
<p
  style={
    Object {
      "color": "red",
    }
  }
>
  YAY
</p>
`;
